src/modules/fancyzones/lib/FancyZones.cpp

```diff
@@ -331,27 +355,25 @@ FancyZones::WindowCreated(HWND window) noexcept
 
     if (m_settings->GetSettings()->appLastZone_moveWindows && IsInterestingWindow(window, m_settings->GetSettings()->excludedAppsArray))
     {
-        auto zoneWindow = m_workAreaHandler.GetWorkArea(window);
-        if (zoneWindow)
+        if (reinterpret_cast<size_t>(::GetProp(window, MULTI_ZONE_STAMP)) == 0)
         {
-            const auto activeZoneSet = zoneWindow->ActiveZoneSet();
-            if (activeZoneSet)
+            POINT cursorPosition{};
+            GUID desktopId{};
+            if (GetCursorPos(&cursorPosition) &&
+                VirtualDesktopUtils::GetWindowDesktopId(window, &desktopId))
             {
-                auto& fancyZonesData = JSONHelpers::FancyZonesDataInstance();
-
-                wil::unique_cotaskmem_string guidString;
-                if (SUCCEEDED(StringFromCLSID(activeZoneSet->Id(), &guidString)))
+                HMONITOR monitor = MonitorFromPoint(cursorPosition, MONITOR_DEFAULTTOPRIMARY);
+                auto zoneWindow = m_workAreaHandler.GetWorkArea(desktopId, monitor);
+                if (zoneWindow && MoveWindowIntoLastKnownPosition(window, zoneWindow))
                 {
-                    std::vector<int> zoneIndexSet = fancyZonesData.GetAppLastZoneIndexSet(window, zoneWindow->UniqueId(), guidString.get());
-                    if (zoneIndexSet.size() &&
-                        !IsSplashScreen(window) &&
-                        !fancyZonesData.IsAnotherWindowOfApplicationInstanceZoned(window, zoneWindow->UniqueId()))
-                    {
-                        m_windowMoveHandler.MoveWindowIntoZoneByIndexSet(window, zoneIndexSet, zoneWindow);
-                        fancyZonesData.UpdateProcessIdToHandleMap(window, zoneWindow->UniqueId());
-                    }
+                    return;
                 }
             }
+            auto zoneWindow = m_workAreaHandler.GetWorkArea(window);
+            if (zoneWindow)
+            {
+                MoveWindowIntoLastKnownPosition(window, zoneWindow);

```

Comment:
Why we call `MoveWindowIntoLastKnownPosition` here?

Reply:
If there isn't a zoneset where the cursor currently is, we fall back to defaults - zone the window to the zoneset of the window's current monitor.

Reply:
This code needs to be refactor because we end up double processing things like `IsSplashScreen()` for no reason.

Reply:
@ivan100sic You're right, this is fallback scenario, in case work are where cursor is has no zone layout, or application has no history there.

Reply:
The function should do all the required processing and then call `MoveWindowIntoZoneByIndexSet` when all the checks are completed.

Reply:
Well, that is only `IsSplashScreen()` as You've mentioned. I'll modify it.

Reply:
@ivan100sic 
the problem is that `MoveWindowIntoLastKnownPosition` returns false also for other reasons other than that, so it's a bad design.

Reply:
I'll move it to the begining of the function, as all other checks are not needed if it's splash screen:
```
// IFancyZonesCallback
IFACEMETHODIMP_(void)
FancyZones::WindowCreated(HWND window) noexcept
{
    std::shared_lock readLock(m_lock);

    if (IsSplashScreen(window))
    {
        return;
    }
```

Reply:
@vldmr11080 
`WindowCreated` should call `MoveWindowIntoLastKnownPosition` only once, when it has a valid `zoneWindow`.

Reply:
That is fine, but that will not change much, we're still gonna call `GetAppLastZoneIndexSet` and `IsAnotherWindowOfApplicationInstanceZoned` for both zone windows. I'll make the change now.

Reply:
Calling `IsAnotherWindowOfApplicationInstanceZoned` twice is not a good design, there is no reason to call it twice, it has to be called only once and only when really needed.

Reply:
Make sure to call `IsSplashScreen()` where `IsInterestingWindow()` is currently called.